solidity/contracts/converter/BancorConverterRegistry.sol

```diff
@@ -76,14 +78,113 @@ contract BancorConverterRegistry is IBancorConverterRegistry, ContractRegistryCl
+            address(converter).transfer(_reserveAmount);
+        }
+
+        token.issue(msg.sender, _reserveAmount);
+        token.transferOwnership(converter);
+        converter.acceptTokenOwnership();
+        converter.transferOwnership(msg.sender);
+
+        addConverter(converter);
+    }
+
+    /**
+      * @dev creates a liquidity-pool converter and adds it to the registry
+      * 
+      * @param _smartTokenName token name
+      * @param _smartTokenSymbol token symbol
+      * @param _smartTokenDecimals token decimals
+      * @param _maxConversionFee maximum conversion-fee
+      * @param _reserveTokens reserve tokens
+      * @param _reserveRatios reserve ratios
+      * @param _reserveAmounts reserve amounts
+    */
+    function newLiquidityPool(
+        string _smartTokenName,
+        string _smartTokenSymbol,
+        uint8 _smartTokenDecimals,
+        uint32 _maxConversionFee,
+        IERC20Token[] memory _reserveTokens,
+        uint32[] memory _reserveRatios,
+        uint256[] memory _reserveAmounts
+    )
+    public payable
+    {
+        uint256 length = _reserveTokens.length;
+        require(length == _reserveRatios.length);
+        require(length == _reserveAmounts.length);
+        require(length > 1);
+
+        IBancorConverterFactory factory = IBancorConverterFactory(addressOf(BANCOR_CONVERTER_FACTORY));
+        SmartToken token = new SmartToken(_smartTokenName, _smartTokenSymbol, _smartTokenDecimals);
+        IBancorConverter converter = IBancorConverter(factory.createConverter(token, registry, _maxConversionFee, IERC20Token(0), 0));
+
+        converter.acceptOwnership();
+
+        for (uint256 i = 0; i < length; i++) {
+            if (_reserveTokens[i] != address(0)) {
+                converter.addReserve(_reserveTokens[i], _reserveRatios[i]);
+                safeTransferFrom(_reserveTokens[i], msg.sender, this, _reserveAmounts[i]);
+                safeApprove(_reserveTokens[i], converter, _reserveAmounts[i]);
+            }
+            else {
+                converter.addETHReserve(_reserveRatios[i]);
+            }
+        }
+
+        token.transferOwnership(converter);
+        converter.acceptTokenOwnership();
+        converter.addLiquidity.value(msg.value)(_reserveTokens, _reserveAmounts, 1);
+        converter.transferOwnership(msg.sender);
+
+        addConverter(converter);
+    }
+
+    /**
+      * @dev adds an existing converter to the registry
+      * only the owner can add an existing converter to the registry, as long as the converter is active and valid
       * note that a liquidity pool converter can be added only if no converter with the same reserve-configuration is already registered
       * 
       * @param _converter converter
     */
-    function addConverter(IBancorConverter _converter) external {
-        // validate input
+    function addConverter(IBancorConverter _converter) public {
+        require(msg.sender == owner || msg.sender == address(this));

```

Comment:
msg.sender == this actually isn't gonna work well since the sender is never address(this).
in addition, we shouldn't block others from calling it until after we gave them a grace period to adapt or all portals that do that will break immediately upon the new deployment

Reply:
@yudilevi : It is when this function is called from function `addLiquidity`.

Reply:
the sender will be the caller to addLiquidity - not the registry

Reply:
@yudilevi : Because we're not "jumping" to another contract?
So how can we prevent non-owners from calling `addConverter`?

Reply:
in this version anyone should be able to call it anyway.
in the next version we should have a shared private function so that the factory functions can call it internally and an owner only public function can call it as well

Reply:
@yudilevi : please see updated code.